controller/volume_controller.go

```diff
@@ -663,7 +663,19 @@ func (vc *VolumeController) hasReplicaEvictionRequested(rs map[string]*longhorn.
 	return false
 }
 
-func (vc *VolumeController) cleanupReplicas(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) error {
+func (vc *VolumeController) cleanupReplicas(v *longhorn.Volume, es map[string]*longhorn.Engine, rs map[string]*longhorn.Replica) error {
+	// TODO: I don't think it's a good idea to cleanup replicas during a migration or engine image update
+	// 	the healthy replica count doesn't differentiate between replicas of different engines
```

Comment:
1. I didn't touch that part, but I agree makes sense to remove it.
2. Agree, already verified the tests. The should have been idempotent to begin with :)
3. MigrationNodeID is not set by the user, it's internal state that is only used by the controller (users only call attach/detach)
4. We can remove this once the UI has been modified, for now this allows for testing. That's why it's marked as HACK for easy removal later.
5. Call attach/detach should stay here, since there is to much overlap and no point having to review the atach/detach twice.

Reply:
Let me provide an example:
If you look at this line, https://github.com/longhorn/longhorn-manager/blob/03a4a77cb6bfbb04323eb0f3b139d4a36821742f/controller/volume_controller.go#L794

Then look at how the healthyReplicaCount is determined:

```go
func (vc *VolumeController) getHealthyReplicaCount(rs map[string]*longhorn.Replica) int {
	var healthyCount int

	for _, r := range rs {
		if r.Spec.FailedAt == "" && r.Spec.HealthyAt != "" {
			healthyCount++
		}
	}

	return healthyCount
}
```

You will see that in the case of a migration/upgrade, this will potentially return upto twice the replicas for the volume.
Which will fulfill this condition `healthyCount > v.Spec.NumberOfReplicas`

Which will then lead to replica deletion, here:
https://github.com/longhorn/longhorn-manager/blob/03a4a77cb6bfbb04323eb0f3b139d4a36821742f/controller/volume_controller.go#L812

Reply:
Can you add this explanation to the comment? A brief version is enough.

Reply:
done